tutorials/rag/agentic_rag.py

```diff
@@ -0,0 +1,274 @@
+
+# Define all the parameters for your pipeline
+MODEL = "gpt-4o-mini"
+# Define your Prompt template that contains variables for answer options required by this example
+# and instructions for tool use
+PROMPT_TEMPLATE = [
+    {
+        "role": "system",
+        "content": """Answer the question provided by the user factually. And reference the options for the answer:
+
+Options:
+- {{option_A}}
+- {{option_B}}
+- {{option_C}}
+- {{option_D}}
+- {{option_E}}
+
+---
+If the question is clearly stated, you can use the `retrieve_knowledge` tool to get more context before answering.
+If you need more information, first call the `get_clarification` tool to ask the user for more details, before then calling the `retrieve_knowledge` tool.
+
+When you are happy with the context, call the `answer_question` tool to provide the answer to the user, where you should provide your reasoning and a clear citation.
+""",
+    }
+]
+
+# Define the tool schemas for the model
+TOOL_SCHEMAS = [
+    {
+        "type": "function",
+        "function": {
+            "name": "retrieve_knowledge",
+            "description": "Looks up relevant context in a knowledge base for a given query.",
+            "parameters": {
+                "type": "object",
+                "required": ["query"],
+                "properties": {
+                    "query": {"type": "string", "description": "The query to retrieve knowledge to help answer."},
+                },
+                "additionalProperties": False,
+            },
+        },
+    },
+    {
+        "type": "function",
+        "function": {
+            "name": "get_clarification",
+            "description": "Asks the user clarifying information.",
+            "parameters": {
+                "type": "object",
+                "required": ["question"],
+                "properties": {
+                    "question": {"type": "string", "description": "Details of the clarification you would like from the user in order to help answer the question."},
+                },
+                "additionalProperties": False,
+            },
+        },
+    },
+    {
+        "type": "function",
+        "function": {
+            "name": "provide_answer",
+            "description": "Submits the answer to the user.",
+            "parameters": {
+                "type": "object",
+                "required": ["answer", "reasoning", "citation"],
+                "properties": {
+                    "answer": {"type": "string", "description": "The final synthesized answer."},
+                    "reasoning": {"type": "string", "description": "The reasoning for your answer."},
+                    "citation": {"type": "string", "description": "The citation to the knowledge base entry you used when answering."},
+                },
+                "additionalProperties": False,
+            },
+        },
+    }
+]
+
+# Implement the tools themselves (these are mock implementations, replace with your own)
```

Comment:
isn't this whole thing a mock implementation? strange that these tools are flagged as such while the prompt etc. isn't.

Reply:
To some extent yes